package app.twallet.air.walletcore.moshi.api

import com.squareup.moshi.JsonClass
import com.squareup.moshi.Types
import org.json.JSONArray
import org.json.JSONObject
import app.twallet.air.walletcontext.models.MBlockchainNetwork
import app.twallet.air.walletcore.api.ArgumentsBuilder
import app.twallet.air.walletcore.models.AccountMfa
import app.twallet.air.walletcore.models.blockchain.MBlockchain
import app.twallet.air.walletcore.moshi.ApiDapp
import app.twallet.air.walletcore.moshi.ApiDappTransfer
import app.twallet.air.walletcore.moshi.inject.ApiDappConnectionRequest
import app.twallet.air.walletcore.moshi.inject.ApiDappDisconnectRequest
import app.twallet.air.walletcore.moshi.inject.ApiDappSessionChain
import app.twallet.air.walletcore.moshi.inject.ApiDappSignDataRequest
import app.twallet.air.walletcore.moshi.inject.ApiDappTransactionRequest
import app.twallet.air.walletcore.moshi.ApiNft
import app.twallet.air.walletcore.moshi.ApiNotificationAddress
import app.twallet.air.walletcore.moshi.ApiSubmitTransferResult
import app.twallet.air.walletcore.moshi.ApiSubmitTransfersResult
import app.twallet.air.walletcore.moshi.ApiTonConnectProof
import app.twallet.air.walletcore.moshi.ApiAddAllFoundSubwalletsResult
import app.twallet.air.walletcore.moshi.ApiAddSubWalletResult
import app.twallet.air.walletcore.moshi.ApiCreateSubWalletResult
import app.twallet.air.walletcore.moshi.ApiGroupedWalletVariant
import app.twallet.air.walletcore.moshi.ApiSubWallet
import app.twallet.air.walletcore.moshi.MApiGetAddressInfoResult
import app.twallet.air.walletcore.moshi.MApiFetchSwapItem
import app.twallet.air.walletcore.moshi.MApiFetchSwapsResult
import app.twallet.air.walletcore.moshi.MApiCheckNftDraftOptions
import app.twallet.air.walletcore.moshi.MApiCheckTransactionDraftOptions
import app.twallet.air.walletcore.moshi.MApiCheckTransactionDraftResult
import app.twallet.air.walletcore.moshi.MApiLedgerAccountInfo
import app.twallet.air.walletcore.moshi.MApiSubmitTransferOptions
import app.twallet.air.walletcore.moshi.MApiSwapEstimateRequest
import app.twallet.air.walletcore.moshi.MApiSwapEstimateResponse
import app.twallet.air.walletcore.moshi.MApiTransaction
import app.twallet.air.walletcore.moshi.MEnvironmentVariables
import app.twallet.air.walletcore.moshi.MImportedViewWalletResponse
import app.twallet.air.walletcore.moshi.MImportedWalletResponse
import app.twallet.air.walletcore.moshi.MRevokeWalletPermissionOptions
import app.twallet.air.walletcore.moshi.MRevokeWalletPermissionResult
import app.twallet.air.walletcore.moshi.MNetworkRpcConfigItem
import app.twallet.air.walletcore.moshi.MRpcResetResult
import app.twallet.air.walletcore.moshi.MRpcTestResult
import app.twallet.air.walletcore.moshi.MRpcUnlockResult
import app.twallet.air.walletcore.moshi.MSignDataPayload
import app.twallet.air.walletcore.moshi.MTonPlugin
import app.twallet.air.walletcore.moshi.MWalletPermission
import app.twallet.air.walletcore.moshi.ReturnStrategy
import app.twallet.air.walletcore.moshi.ledger.MLedgerWalletInfo
import java.lang.reflect.Type
import java.math.BigInteger

sealed class ApiMethod<T> {
    abstract val name: String
    abstract val type: Type
    abstract val arguments: String

    /* Other */
    object Other {
        class SetIsAppFocused(
            isFocused: Boolean
        ) : ApiMethod<Array<String>>() {
            override val name: String = "setIsAppFocused"
            override val type: Type = Any::class.java
            override val arguments: String = ArgumentsBuilder()
                .boolean(isFocused)
                .build()
        }

        class WaitForLedgerApp(chain: MBlockchain, options: Options? = null) :
            ApiMethod<Boolean>() {
            @JsonClass(generateAdapter = true)
            data class Options(
                val timeout: Int? = null,
                val attemptPause: Int? = null
            )

            override val name: String = "waitForLedgerApp"
            override val type: Type = Boolean::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .jsObject(options, Options::class.java)
                .build()
        }

        class RenderBlurredReceiveBg(
            chain: MBlockchain,
            options: Options? = null
        ) : ApiMethod<String>() {
            @JsonClass(generateAdapter = true)
            data class Options(
                val width: Int? = null,
                val height: Int? = null,
                val blurPx: Int? = null,
                val quality: Double? = null,
                val overlay: String? = null,
                val scale: Int? = null
            )

            override val name = "renderBlurredReceiveBg"
            override val type: Type = String::class.java
            override val arguments = ArgumentsBuilder()
                .string(chain.name)
                .jsObject(options, Options::class.java)
                .build()
        }

        class GetEnvironmentVariables : ApiMethod<MEnvironmentVariables>() {
            override val name: String = "getEnvironmentVariables"
            override val type: Type = MEnvironmentVariables::class.java
            override val arguments: String = ArgumentsBuilder()
                .build()
        }
    }

    /* Auth */
    object Auth {
        class GenerateMnemonic(
            isBip39: Boolean = true,
        ) : ApiMethod<Array<String>>() {
            override val name: String = "generateMnemonic"
            override val type: Type = Array<String>::class.java
            override val arguments: String = ArgumentsBuilder()
                .boolean(isBip39)
                .build()
        }

        class GetLedgerWallets(
            chain: MBlockchain,
            network: MBlockchainNetwork,
            startWalletIndex: Int,
            count: Int
        ) : ApiMethod<Array<MLedgerWalletInfo>>() {
            override val name: String = "getLedgerWallets"
            override val type: Type = Array<MLedgerWalletInfo>::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .string(network.value)
                .number(startWalletIndex)
                .number(count)
                .build()
        }

        class ImportLedgerWallet(
            network: MBlockchainNetwork,
            accountInfo: MApiLedgerAccountInfo
        ) : ApiMethod<MImportedWalletResponse>() {
            override val name: String = "importLedgerAccount"
            override val type: Type = MImportedWalletResponse::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(network.value)
                .jsObject(accountInfo, MApiLedgerAccountInfo::class.java)
                .build()
        }

        class ImportViewAccount(
            network: MBlockchainNetwork,
            addressByChain: Map<MBlockchain, String>
        ) : ApiMethod<MImportedViewWalletResponse>() {
            override val name: String = "importViewAccount"
            override val type: Type = MImportedViewWalletResponse::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(network.value)
                .jsonObject(JSONObject().apply {
                    addressByChain.forEach { (chain, address) ->
                        put(chain.name, address)
                    }
                })
                .build()
        }
    }

    /* Wallet Data */
    object WalletData {
        class GetAddressInfo(
            chain: MBlockchain,
            network: MBlockchainNetwork,
            addressOrDomain: String,
        ) : ApiMethod<MApiGetAddressInfoResult>() {
            override val name: String = "getAddressInfo"
            override val type: Type = MApiGetAddressInfoResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .string(network.value)
                .string(addressOrDomain)
                .build()
        }

        class DecryptComment(
            accountId: String,
            activity: MApiTransaction,
            passcode: String
        ) : ApiMethod<String>() {
            override val name: String = "decryptComment"
            override val type: Type = String::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsObject(activity, MApiTransaction::class.java)
                .string(passcode)
                .build()
        }

        class FetchActivityDetails(
            accountId: String,
            activity: MApiTransaction,
        ) : ApiMethod<MApiTransaction>() {
            override val name: String = "fetchActivityDetails"
            override val type: Type = MApiTransaction::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsObject(activity, MApiTransaction::class.java)
                .build()
        }

        class FetchPastActivities(
            accountId: String,
            limit: Int,
            slug: String?,
            toTimestamp: Long?,
        ) : ApiMethod<FetchPastActivities.Result>() {

            @JsonClass(generateAdapter = true)
            data class Result(
                val activities: List<MApiTransaction>,
                val hasMore: Boolean
            )

            override val name: String = "fetchPastActivities"
            override val type: Type = Result::class.java
            override val arguments: String = run {
                var builder = ArgumentsBuilder()
                    .string(accountId)
                    .number(limit)
                    .string(slug)

                toTimestamp?.let {
                    builder = builder.number(it)
                }

                builder.build()
            }
        }

        class FetchTransactionById(
            options: Options
        ) : ApiMethod<List<MApiTransaction>>() {

            @JsonClass(generateAdapter = true)
            data class Options(
                val chain: String,
                val network: String,
                val walletAddress: String,
                val txId: String? = null,
                val txHash: String? = null
            )

            override val name: String = "fetchTransactionById"
            override val type: Type =
                Types.newParameterizedType(List::class.java, MApiTransaction::class.java)
            override val arguments: String = ArgumentsBuilder()
                .jsObject(options, Options::class.java)
                .build()
        }
    }

    /* Tokens */
    object Tokens {
        class BuildTokenSlug(
            chain: String,
            address: String,
        ) : ApiMethod<String>() {
            override val name: String = "buildTokenSlug"
            override val type: Type = String::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain)
                .string(address)
                .build()
        }
    }

    object Settings {
        class FetchMnemonic(
            accountId: String,
            password: String
        ) : ApiMethod<Array<String>>() {
            override val name: String = "fetchMnemonic"
            override val type: Type = Array<String>::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(password)
                .build()
        }

        class ChangePassword(
            oldPasscode: String,
            newPasscode: String
        ) : ApiMethod<Nothing>() {
            override val name: String = "changePassword"
            override val type: Type = Nothing::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(oldPasscode)
                .string(newPasscode)
                .build()
        }

        class GetWalletVariants(
            accountId: String,
            page: Int,
            mnemonic: Array<String>
        ) : ApiMethod<Array<ApiGroupedWalletVariant>>() {
            override val name: String = "getWalletVariants"
            override val type: Type = Array<ApiGroupedWalletVariant>::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .number(page)
                .jsArray(mnemonic.toList(), String::class.java)
                .build()
        }

        class CreateSubWallet(
            accountId: String,
            password: String
        ) : ApiMethod<ApiCreateSubWalletResult>() {
            override val name: String = "createSubWallet"
            override val type: Type = ApiCreateSubWalletResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(password)
                .build()
        }

        class AddSubWallet(
            accountId: String,
            byChain: Map<String, ApiSubWallet>
        ) : ApiMethod<ApiAddSubWalletResult>() {
            override val name: String = "addSubWallet"
            override val type: Type = ApiAddSubWalletResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsObject(
                    byChain,
                    Types.newParameterizedType(
                        Map::class.java,
                        String::class.java,
                        ApiSubWallet::class.java
                    )
                )
                .build()
        }

        class AddAllFoundSubwallets(
            accountId: String,
            foundWallets: List<Map<String, ApiSubWallet>>
        ) : ApiMethod<ApiAddAllFoundSubwalletsResult>() {
            override val name: String = "addAllFoundSubwallets"
            override val type: Type = ApiAddAllFoundSubwalletsResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsArray(
                    foundWallets,
                    Types.newParameterizedType(
                        Map::class.java,
                        String::class.java,
                        ApiSubWallet::class.java
                    )
                )
                .build()
        }
    }

    /* Transfer */

    object Transfer {
        class CheckTransactionDraft(
            chain: MBlockchain,
            options: MApiCheckTransactionDraftOptions
        ) : ApiMethod<MApiCheckTransactionDraftResult>() {
            override val name: String = "checkTransactionDraft"
            override val type: Type = MApiCheckTransactionDraftResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .jsObject(options, MApiCheckTransactionDraftOptions::class.java)
                .build()
        }

        class SubmitTransfer(
            chain: MBlockchain,
            options: MApiSubmitTransferOptions
        ) : ApiMethod<ApiSubmitTransferResult>() {
            override val name: String = "submitTransfer"
            override val type: Type = ApiSubmitTransferResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .jsObject(options, MApiSubmitTransferOptions::class.java)
                .build()
        }

        class SignDappTransfers(
            dappChain: ApiDappSessionChain,
            accountId: String,
            transactions: List<ApiDappTransfer>,
            options: Options
        ) : ApiMethod<Any>() {

            @JsonClass(generateAdapter = true)
            data class Options(
                val password: String?,
                val validUntil: Long?,
                val vestingAddress: String?,
                val isLegacyOutput: Boolean?
            )

            override val name: String = "signDappTransfers"
            override val type: Type = Any::class.java
            override val arguments: String = ArgumentsBuilder()
                .jsObject(dappChain, ApiDappSessionChain::class.java)
                .string(accountId)
                .jsArray(transactions, ApiDappTransfer::class.java)
                .jsObject(options, Options::class.java)
                .build()
        }

        class SignDappData(
            dappChain: ApiDappSessionChain,
            accountId: String,
            dappUrl: String,
            payloadToSign: MSignDataPayload,
            password: String,
        ) : ApiMethod<JSONObject>() {

            override val name: String = "signDappData"
            override val type: Type = JSONObject::class.java
            override val arguments: String = ArgumentsBuilder()
                .jsObject(dappChain, ApiDappSessionChain::class.java)
                .string(accountId)
                .string(dappUrl)
                .jsObject(payloadToSign, MSignDataPayload::class.java)
                .string(password)
                .build()
        }
    }


    /* Swap */

    object Swap {
        class SwapEstimate(
            accountId: String,
            request: MApiSwapEstimateRequest
        ) : ApiMethod<MApiSwapEstimateResponse>() {
            override val name: String = "swapEstimate"
            override val type: Type = MApiSwapEstimateResponse::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsObject(request, MApiSwapEstimateRequest::class.java)
                .build()
        }

        class FetchSwaps(
            accountId: String,
            items: List<MApiFetchSwapItem>
        ) : ApiMethod<MApiFetchSwapsResult>() {
            override val name: String = "fetchSwaps"
            override val type: Type = MApiFetchSwapsResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsArray(items, MApiFetchSwapItem::class.java)
                .build()
        }

        class ConfirmSwapMfaRequest(
            accountId: String,
            swapId: String,
            txHash: String,
        ) : ApiMethod<Unit>() {
            override val name: String = "confirmSwapMfaRequest"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(swapId)
                .string(txHash)
                .build()
        }
    }


    /* Ton Connect */

    object DApp {
        class GetDapps(
            accountId: String
        ) : ApiMethod<List<ApiDapp>>() {
            override val name: String = "getDapps"
            override val type: Type =
                Types.newParameterizedType(List::class.java, ApiDapp::class.java)
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .build()
        }

        class TonConnectHandleDeepLink(
            url: String,
            isFromInAppBrowser: Boolean? = null,
            identifier: String? = null
        ) : ApiMethod<ReturnStrategy?>() {
            override val name: String = "tonConnect_handleDeepLink"
            override val type: Type = ReturnStrategy::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(url)
                .boolean(isFromInAppBrowser)
                .string(identifier)
                .build()
        }

        class ConfirmDappRequestSendTransaction(
            promiseId: String,
            signedMessages: JSONArray
        ) : ApiMethod<Unit>() {
            override val name: String = "confirmDappRequestSendTransaction"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .jsObject(signedMessages, JSONArray::class.java)
                .build()
        }

        class ConfirmDappRequestSendTransactionMfa(
            promiseId: String,
            mfaRequestHash: String,
        ) : ApiMethod<Unit>() {
            @JsonClass(generateAdapter = true)
            data class Payload(val mfaRequestHash: String)

            override val name: String = "confirmDappRequestSendTransaction"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .jsObject(Payload(mfaRequestHash), Payload::class.java)
                .build()
        }

        class ConfirmDappRequestSignData(
            promiseId: String,
            signedData: JSONObject
        ) : ApiMethod<Unit>() {
            override val name: String = "confirmDappRequestSignData"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .jsObject(signedData, JSONObject::class.java)
                .build()
        }

        class SignDappProof(
            dappChains: List<ApiDappSessionChain>,
            accountId: String,
            proofData: ApiTonConnectProof?,
            password: String
        ) : ApiMethod<SignDappProof.Result>() {

            @JsonClass(generateAdapter = true)
            data class Result(val signatures: List<String>)

            override val name: String = "signDappProof"
            override val type: Type = Result::class.java
            override val arguments: String = ArgumentsBuilder()
                .jsArray(dappChains, ApiDappSessionChain::class.java)
                .string(accountId)
                .jsObject(proofData, ApiTonConnectProof::class.java)
                .string(password)
                .build()
        }

        class ConfirmDappRequestConnect(
            promiseId: String,
            request: Request
        ) : ApiMethod<Unit>() {
            override val name: String = "confirmDappRequestConnect"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .jsObject(request, Request::class.java)
                .build()

            @JsonClass(generateAdapter = true)
            data class Request(
                val accountId: String? = null,
                val proofSignatures: List<String>? = null
            )
        }

        class CreateDappConnectMfaRequest(
            accountId: String,
            password: String?,
        ) : ApiMethod<CreateDappConnectMfaRequest.Response>() {
            override val name: String = "createDappConnectMfaRequest"
            override val type: Type = Response::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .apply { password?.let { string(it) } }
                .build()

            @JsonClass(generateAdapter = true)
            data class Response(
                val mfaRequestHash: String? = null,
                val error: String? = null,
            )
        }

        class CancelDappRequest(
            promiseId: String,
            reason: String?
        ) : ApiMethod<Unit>() {
            override val name: String = "cancelDappRequest"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .apply { reason?.let { string(it) } }
                .build()
        }

        class RecordTonConnectEvent(
            eventName: String,
            promiseId: String
        ) : ApiMethod<Unit>() {
            override val name: String = "recordTonConnectEvent"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .jsonObject(JSONObject().apply {
                    put("event_name", eventName)
                    put("promiseId", promiseId)
                })
                .build()
        }

        class DeleteDapp(
            accountId: String,
            appClientId: String,
            origin: String
        ) : ApiMethod<Any>() {
            override val name: String = "deleteDapp"
            override val type: Type = Any::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(origin)
                .string(appClientId)
                .string(null)
                .build()
        }

        class DeleteAllDapps(
            accountId: String
        ) : ApiMethod<Boolean>() {
            override val name: String = "deleteAllDapps"
            override val type: Type = Boolean::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .build()
        }


        object Inject {
            @JsonClass(generateAdapter = true)
            data class DAppArg(
                val url: String,
                val urlTrustStatus: String = "verified",
                var accountId: String,
            )

            class TonConnectConnect(
                dApp: DAppArg,
                request: ApiDappConnectionRequest,
                requestId: Int
            ) : ApiMethod<JSONObject>() {
                override val name: String = "tonConnect_connect"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, ApiDappConnectionRequest::class.java)
                    .number(requestId)
                    .build()
            }

            class TonConnectReconnect(
                dApp: DAppArg,
                requestId: Int
            ) : ApiMethod<JSONObject>() {
                override val name: String = "tonConnect_reconnect"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .number(requestId)
                    .build()
            }

            class TonConnectDisconnect(
                dApp: DAppArg,
                request: ApiDappDisconnectRequest
            ) : ApiMethod<JSONObject>() {
                override val name: String = "tonConnect_disconnect"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, ApiDappDisconnectRequest::class.java)
                    .build()
            }

            class TonConnectSendTransaction(
                dApp: DAppArg,
                request: ApiDappTransactionRequest
            ) : ApiMethod<JSONObject>() {
                override val name: String = "tonConnect_sendTransaction"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, ApiDappTransactionRequest::class.java)
                    .build()
            }

            class TonConnectSignData(
                dApp: DAppArg,
                request: ApiDappSignDataRequest
            ) : ApiMethod<JSONObject>() {
                override val name: String = "tonConnect_signData"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, ApiDappSignDataRequest::class.java)
                    .build()
            }

            class WalletConnectConnect(
                dApp: DAppArg,
                request: Any,
                requestId: Int
            ) : ApiMethod<JSONObject>() {
                override val name: String = "walletConnect_connect"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, Any::class.java)
                    .number(requestId)
                    .build()
            }

            class WalletConnectReconnect(
                dApp: DAppArg,
                requestId: Int
            ) : ApiMethod<JSONObject>() {
                override val name: String = "walletConnect_reconnect"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .number(requestId)
                    .build()
            }

            class WalletConnectDisconnect(
                dApp: DAppArg,
                request: Any
            ) : ApiMethod<JSONObject>() {
                override val name: String = "walletConnect_disconnect"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, Any::class.java)
                    .build()
            }

            class WalletConnectSendTransaction(
                dApp: DAppArg,
                request: Any
            ) : ApiMethod<JSONObject>() {
                override val name: String = "walletConnect_sendTransaction"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, Any::class.java)
                    .build()
            }

            class WalletConnectSignData(
                dApp: DAppArg,
                request: Any
            ) : ApiMethod<JSONObject>() {
                override val name: String = "walletConnect_signData"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, Any::class.java)
                    .build()
            }

            class WalletConnectProxyEvmRpc(
                dApp: DAppArg,
                request: Any
            ) : ApiMethod<JSONObject>() {
                override val name: String = "walletConnect_proxyEvmRpc"
                override val type: Type = JSONObject::class.java
                override val arguments: String = ArgumentsBuilder()
                    .jsObject(dApp, DAppArg::class.java)
                    .jsObject(request, Any::class.java)
                    .build()
            }
        }

        class WalletConnectHandleDeepLink(
            url: String
        ) : ApiMethod<Any?>() {
            override val name: String = "walletConnect_handleDeepLink"
            override val type: Type = Any::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(url)
                .build()
        }

        class ConfirmWalletConnectPaySignTransaction(
            promiseId: String,
            signedTransactions: JSONArray
        ) : ApiMethod<Unit>() {
            override val name: String = "confirmWalletConnectPaySignTransaction"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .jsObject(signedTransactions, JSONArray::class.java)
                .build()
        }

        class ConfirmWalletConnectPaySignData(
            promiseId: String,
            signedData: JSONObject
        ) : ApiMethod<Unit>() {
            override val name: String = "confirmWalletConnectPaySignData"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .jsObject(signedData, JSONObject::class.java)
                .build()
        }

        class CompleteWalletConnectPayDataCollection(
            promiseId: String
        ) : ApiMethod<Unit>() {
            override val name: String = "completeWalletConnectPayDataCollection"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .build()
        }

        class ConfirmWalletConnectPayOptionSelection(
            promiseId: String,
            optionId: String
        ) : ApiMethod<Unit>() {
            override val name: String = "confirmWalletConnectPayOptionSelection"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .string(optionId)
                .build()
        }

        class CancelWalletConnectPay(
            promiseId: String,
            reason: String?
        ) : ApiMethod<Unit>() {
            override val name: String = "cancelWalletConnectPay"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(promiseId)
                .apply { reason?.let { string(it) } }
                .build()
        }

        class RefreshWalletConnectPayOptionSelection(
            paymentLink: String,
            accountId: String,
            promiseId: String
        ) : ApiMethod<Unit>() {
            override val name: String = "refreshWalletConnectPayOptionSelection"
            override val type: Type = Unit::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(paymentLink)
                .string(accountId)
                .string(promiseId)
                .build()
        }
    }

    /* Domains */
    object Domains {
        class CheckDnsRenewalDraft(accountId: String, nfts: List<ApiNft>) :
            ApiMethod<CheckDnsRenewalDraft.Result>() {

            @JsonClass(generateAdapter = true)
            data class Result(val realFee: BigInteger)

            override val name: String = "checkDnsRenewalDraft"
            override val type: Type = Result::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsArray(nfts, ApiNft::class.java)
                .build()
        }

        class SubmitDnsRenewal(
            accountId: String,
            password: String,
            nfts: List<ApiNft>,
            realFee: BigInteger
        ) : ApiMethod<ApiSubmitTransferResult>() {
            override val name: String = "submitDnsRenewal"
            override val type: Type = ApiSubmitTransferResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(password)
                .jsArray(nfts, ApiNft::class.java)
                .bigInt(realFee)
                .build()
        }

        class CheckDnsChangeWalletDraft(
            accountId: String,
            nft: ApiNft,
            address: String
        ) : ApiMethod<CheckDnsChangeWalletDraft.Result>() {

            @JsonClass(generateAdapter = true)
            data class Result(val realFee: BigInteger)

            override val name: String = "checkDnsChangeWalletDraft"
            override val type: Type = Result::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsObject(nft, ApiNft::class.java)
                .string(address)
                .build()
        }

        class SubmitDnsChangeWallet(
            accountId: String,
            password: String,
            nft: ApiNft,
            address: String,
            realFee: BigInteger,
        ) : ApiMethod<ApiSubmitTransferResult>() {

            override val name: String = "submitDnsChangeWallet"
            override val type: Type = ApiSubmitTransferResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(password)
                .jsObject(nft, ApiNft::class.java)
                .string(address)
                .bigInt(realFee)
                .build()
        }
    }

    /* Nft */
    object Nft {
        class FetchNftByAddress(
            network: MBlockchainNetwork,
            nftAddress: String,
        ) : ApiMethod<ApiNft?>() {
            override val name: String = "fetchNftByAddress"
            override val type: Type = ApiNft::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(network.value)
                .string(nftAddress)
                .build()
        }

        class CheckNftTransferDraft(
            chain: MBlockchain,
            options: MApiCheckNftDraftOptions
        ) : ApiMethod<MApiCheckTransactionDraftResult>() {
            override val name: String = "checkNftTransferDraft"
            override val type: Type = MApiCheckTransactionDraftResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .jsObject(options, MApiCheckNftDraftOptions::class.java)
                .build()
        }

        class SubmitNftTransfer(
            chain: MBlockchain,
            accountId: String,
            passcode: String,
            nfts: List<ApiNft>,
            address: String,
            comment: String?,
            fee: BigInteger,
            isNftBurn: Boolean,
            addressName: String? = null,
        ) : ApiMethod<ApiSubmitTransfersResult>() {
            override val name: String = "submitNftTransfers"
            override val type: Type = ApiSubmitTransfersResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .string(accountId)
                .string(passcode)
                .jsObject(
                    nfts.map { it.toDictionary() }.toTypedArray(),
                    Array<JSONObject>::class.java
                )
                .string(address)
                .string(comment)
                .bigInt(fee)
                .boolean(isNftBurn)
                .string(addressName)
                .build()
        }

        class CheckNftOwnership(
            chain: String,
            accountId: String,
            nftAddress: String,
        ) : ApiMethod<Any>() {
            override val name: String = "checkNftOwnership"
            override val type: Type = Boolean::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain)
                .string(accountId)
                .string(nftAddress)
                .build()
        }

        class FetchNftsFromCollection(
            accountId: String,
            collection: Collection,
        ) : ApiMethod<Any>() {

            @JsonClass(generateAdapter = true)
            data class Collection(
                val chain: String,
                val address: String,
            )

            override val name: String = "fetchNftsFromCollection"
            override val type: Type = Boolean::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsObject(collection, Collection::class.java)
                .build()
        }

        /** Start/stop background NFT scanning (Collectibles tab visibility). */
        class SetCollectiblesActive(
            isActive: Boolean,
        ) : ApiMethod<MRpcResetResult>() {
            override val name: String = "setCollectiblesActive"
            override val type: Type = MRpcResetResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .boolean(isActive)
                .build()
        }
    }


    /* Notifications */
    object Notifications {
        class SubscribeNotifications(props: Props) : ApiMethod<JSONObject>() {

            @JsonClass(generateAdapter = true)
            data class Props(
                val userToken: String,
                val addresses: List<ApiNotificationAddress>,
                val langCode: String,
                val platform: String = "android",
            )

            override val name: String = "subscribeNotifications"
            override val type: Type = JSONObject::class.java
            override val arguments: String = ArgumentsBuilder()
                .jsObject(props, Props::class.java)
                .build()
        }

        class UnsubscribeNotifications(props: Props) : ApiMethod<Any>() {

            @JsonClass(generateAdapter = true)
            data class Props(
                val userToken: String,
                val addresses: List<ApiNotificationAddress>
            )

            override val name: String = "unsubscribeNotifications"
            override val type: Type = Any::class.java
            override val arguments: String = ArgumentsBuilder()
                .jsObject(props, Props::class.java)
                .build()
        }
    }

    /* MFA */
    object Mfa {
        @JsonClass(generateAdapter = true)
        data class ApiMfaRequestCreated(
            val reqId: String,
        )

        class FetchMfaRequest(hash: String) : ApiMethod<FetchMfaRequest.ApiMfaRequest>() {
            @JsonClass(generateAdapter = true)
            data class ApiMfaRequest(
                val payload: String,
                val signature: String,
                val isConfirmed: Boolean,
                val txHash: String,
            )

            override val name: String = "fetchMfaRequest"
            override val type: Type = ApiMfaRequest::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(hash)
                .build()
        }

        class FetchInstallMfaRequest(reqId: String) :
            ApiMethod<FetchInstallMfaRequest.ApiInstallMfaRequest>() {
            @JsonClass(generateAdapter = true)
            data class ApiInstallMfaRequest(
                val address: String,
                val user: AccountMfa.User? = null,
            )

            override val name: String = "fetchInstallMfaRequest"
            override val type: Type = ApiInstallMfaRequest::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(reqId)
                .build()
        }

        class PublishInstallMfaRequest(accountId: String) : ApiMethod<ApiMfaRequestCreated>() {
            override val name: String = "publishInstallMfaRequest"
            override val type: Type = ApiMfaRequestCreated::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .build()
        }

        class InstallMfaFromRequest(
            accountId: String,
            user: AccountMfa.User,
            password: String?,
        ) : ApiMethod<String>() {
            override val name: String = "installMfaFromRequest"
            override val type: Type = String::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .jsObject(
                    user,
                    AccountMfa.User::class.java
                )
                .string(password)
                .build()
        }

        class PublishRemoveMfaRequest(
            accountId: String,
            password: String?,
        ) : ApiMethod<ApiMfaRequestCreated>() {
            override val name: String = "publishRemoveMfaRequest"
            override val type: Type = ApiMfaRequestCreated::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(password)
                .build()
        }

        class ConfirmMfaRemovalRequest(accountId: String) : ApiMethod<Any>() {
            override val name: String = "confirmMfaRemovalRequest"
            override val type: Type = Any::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .build()
        }

        class RefreshMfaState(
            accountId: String,
            password: String?,
        ) : ApiMethod<RefreshMfaState.Response>() {
            @JsonClass(generateAdapter = true)
            data class Response(
                val changed: Boolean = false,
                val mfa: AccountMfa? = null,
            )

            override val name: String = "refreshMfaState"
            override val type: Type = Response::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(password)
                .build()
        }
    }

    /* Permissions */

    object Permissions {
        class FetchWalletPermissions(
            accountId: String,
            chain: MBlockchain,
        ) : ApiMethod<List<MWalletPermission>>() {
            override val name: String = "fetchWalletPermissions"
            override val type: Type =
                Types.newParameterizedType(List::class.java, MWalletPermission::class.java)
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .string(chain.name)
                .build()
        }

        class FetchWalletPlugins(
            accountId: String,
        ) : ApiMethod<List<MTonPlugin>>() {
            override val name: String = "fetchWalletPlugins"
            override val type: Type =
                Types.newParameterizedType(List::class.java, MTonPlugin::class.java)
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .build()
        }

        class RevokeWalletPermission(
            chain: MBlockchain,
            options: MRevokeWalletPermissionOptions,
        ) : ApiMethod<MRevokeWalletPermissionResult>() {
            override val name: String = "revokeWalletPermission"
            override val type: Type = MRevokeWalletPermissionResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain.name)
                .apply {
                    when (options) {
                        is MRevokeWalletPermissionOptions.Approval ->
                            jsObject(options, MRevokeWalletPermissionOptions.Approval::class.java)

                        is MRevokeWalletPermissionOptions.Delegation ->
                            jsObject(options, MRevokeWalletPermissionOptions.Delegation::class.java)
                    }
                }
                .build()
        }
    }

    object Networks {
        class GetRpcConfig(
            network: String,
        ) : ApiMethod<Array<MNetworkRpcConfigItem>>() {
            override val name: String = "getRpcConfig"
            override val type: Type = Array<MNetworkRpcConfigItem>::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(network)
                .build()
        }

        class TestRpcEndpoint(
            chain: String,
            network: String,
            field: String,
            url: String,
            apiKey: String? = null,
        ) : ApiMethod<MRpcTestResult>() {
            override val name: String = "testRpcEndpoint"
            override val type: Type = MRpcTestResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain)
                .string(network)
                .string(field)
                .string(url)
                .string(apiKey)
                .build()
        }

        class SetRpcOverride(
            chain: String,
            network: String,
            field: String,
            url: String,
            apiKey: String? = null,
            force: Boolean? = null,
            password: String? = null,
        ) : ApiMethod<MRpcTestResult>() {
            override val name: String = "setRpcOverride"
            override val type: Type = MRpcTestResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain)
                .string(network)
                .string(field)
                .string(url)
                .string(apiKey)
                .boolean(force)
                .string(password)
                .build()
        }

        class ResetRpcOverride(
            chain: String,
            network: String,
            field: String,
        ) : ApiMethod<MRpcResetResult>() {
            override val name: String = "resetRpcOverride"
            override val type: Type = MRpcResetResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain)
                .string(network)
                .string(field)
                .build()
        }

        class UnlockRpcApiKey(
            chain: String,
            network: String,
            password: String,
            field: String = "rpc",
        ) : ApiMethod<MRpcUnlockResult>() {
            override val name: String = "unlockRpcApiKey"
            override val type: Type = MRpcUnlockResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain)
                .string(network)
                .string(password)
                .string(field)
                .build()
        }

        class SetChainVisibility(
            chain: String,
            network: String,
            isHidden: Boolean,
        ) : ApiMethod<MRpcResetResult>() {
            override val name: String = "setChainVisibility"
            override val type: Type = MRpcResetResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(chain)
                .string(network)
                .boolean(isHidden)
                .build()
        }

        class SetAccountVaultProfile(
            accountId: String,
            isVault: Boolean,
        ) : ApiMethod<MRpcResetResult>() {
            override val name: String = "setAccountVaultProfile"
            override val type: Type = MRpcResetResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .string(accountId)
                .boolean(isVault)
                .build()
        }

        class SyncVaultAccounts(
            accountIds: List<String>,
        ) : ApiMethod<MRpcResetResult>() {
            override val name: String = "syncVaultAccounts"
            override val type: Type = MRpcResetResult::class.java
            override val arguments: String = ArgumentsBuilder()
                .jsArray(accountIds, String::class.java)
                .build()
        }
    }
}
